tpe/plugin/src/Base.hh

```diff
@@ -88,26 +88,38 @@ class Base : public Implements3d<FeatureList<Feature>>
     return -1;
   }
 
-  public: inline std::size_t indexInContainerToId(
-    const std::size_t _containerId, const std::size_t _index) const
+  public: template <typename EntityType>
+  inline std::pair<std::size_t, EntityType> indexInContainerToId(
+      const std::size_t _containerId,
+      const std::size_t _index,
+      const std::map<std::size_t, EntityType> &_idMap) const
   {
     std::size_t counter = 0;
     auto it = this->childIdToParentId.begin();
 
     while (counter <= _index && it != this->childIdToParentId.end())
     {
-      if (it->second == _containerId && counter == _index)
+      if (it->second == _containerId)
       {
-        return it->first;
-      }
-      else if (it->second == _containerId)
-      {
-        ++counter;
+        auto idMapIt = _idMap.find(it->first);
+        // only count if the entity is found in the idMap. This makes sure we're
+        // counting only the entities with the correct EntityType
+        if (idMapIt != _idMap.end())
+        {
+          if (counter == _index)
```

Comment:
I'm having a hard time wrapping my mind around the purpose of the `counter` variable in this method, would you mind helping me understand why we need it?

Reply:
Sure. `childIdToParentId`, is a map from child Entity IDs to parent Entity IDs. These are IDs, not indices. But we are trying to look up an index in the container identified by `_containerId`. To do so, we have to go through all the key-value pairs in `childToParentId` and count the elements whose values equal `_containerId` and whose keys are found in `_idMap`. `_idMap` is a container of specific types of entities. We need it because `_containerId` doesn't differentiate between links and models. For example, if we want to find the second link in a model, `_idMap` would be `Base::links`, but if we wanted to find the second model in a world `_idMap` would be `Base::models`.

Counting the elements in `childIdToParentId` works because it's an ordered map and the entity IDs of all the entity types are created in a globally increasing sequence. So, when iterating through `childIdToParentId`, we get entities in the order of creation.

Reply:
> Counting the elements in childIdToParentId works because it's an ordered map

Ah, I think this was the key piece I was missing - I glanced over the fact that we are using `std::map` and not something like `std::unordered_map`. Thanks for the explanation, that helps a lot!